src/proof/er/er_proof.cpp

```diff
@@ -0,0 +1,368 @@
+                               std::move(otherLiterals));
+
+    // Advance over the lines for this definition
+    i += nLinesForThisDef;
+  }
+}
+
+void ErProof::outputAsLfsc(std::ostream& os) const
+{
+  // How many parens to close?
+  size_t parenCount = 0;
+  std::unordered_set<prop::SatVariable> newVariables;
+
+  // Print Definitions
+  for (const ErDefinition& def : d_definitions)
+  {
+    os << "\n    (decl_rat_elimination_def ("
+       << (def.d_oldLiteral.isNegated() ? "neg " : "pos ")
+       << ProofManager::getVarName(def.d_oldLiteral.getSatVariable(), "bb")
+       << ") ";
+    LFSCProofPrinter::printSatClause(def.d_otherLiterals, os, "bb");
+    os << " (\\ er.v" << def.d_newVariable << " (\\ er.def"
+       << def.d_newVariable;
+    newVariables.insert(def.d_newVariable);
+  }
+  parenCount += 3 * d_definitions.size();
+
+  // Clausify Definitions
+  TraceCheckIdx firstDefClause = d_inputClauseIds.size() + 1;
+  for (const ErDefinition& def : d_definitions)
+  {
+    os << "\n    (clausify_rat_elimination_def _ _ _ "
+       << "er.def " << def.d_newVariable << " _ _ (\\ er.c" << firstDefClause
+       << " (\\ er.c" << (firstDefClause + 1) << " (\\ er.cnf"
+       << def.d_newVariable;
+
+    firstDefClause += 2 + def.d_otherLiterals.size();
+  }
+  parenCount += 4 * d_definitions.size();
+
+  // Unroll proofs of CNFs to proofs of clauses
+  firstDefClause = d_inputClauseIds.size() + 1;
+  for (const ErDefinition& def : d_definitions)
+  {
+    for (size_t i = 0, n = def.d_otherLiterals.size(); i < n; ++i)
+    {
+      os << "\n    (cnfc_unroll _ _ ";
+      os << "er.cnf" << def.d_newVariable;
+      if (i != 0)
+      {
+        os << ".u" << i;
+      }
+      os << " (\\ er.c" << (firstDefClause + 2 + i);
+      os << " (\\ er.cnf" << def.d_newVariable << ".u" << (i + 1);
+    }
+    parenCount += 3 * def.d_otherLiterals.size();
+
+    firstDefClause += 2 + def.d_otherLiterals.size();
+  }
+
+  // NB: At this point `firstDefClause` points to the first clause resulting
+  // from a resolution chain
+
+  // Now, elaborate each resolution chain
+  for (size_t cId = firstDefClause, nLines = d_tracecheck.d_lines.size();
+       cId <= nLines;
+       ++cId)
+  {
+    const std::vector<TraceCheckIdx>& chain =
+        d_tracecheck.d_lines[cId - 1].d_chain;
+    const std::vector<prop::SatLiteral> pivots = computePivotsForChain(chain);
+    Assert(chain.size() > 0);
+    Assert(chain.size() == pivots.size() + 1);
+
+    os << "\n    (satlem_simplify _ _ _ ";
+    parenCount += 1;
+
+    // Print resolution openings (reverse order)
+    for (int64_t i = pivots.size() - 1; i >= 0; --i)
+    {
+      prop::SatLiteral pivot = pivots[i];
+      os << "(" << (pivot.isNegated() ? 'Q' : 'R') << " _ _ ";
+    }
+
+    // Print resolution start
+    writeIdForClauseProof(os, chain[0]);
+    os << " ";
+
+    // Print resolution closings (forward order)
+    for (size_t i = 0, n = pivots.size(); i < n; ++i)
```

Comment:
Maybe use `for (auto pivot : pivots) { ... }`

Reply:
Alas, we need `i` within the loop in order to index `chain` properly. If C++ had a zip, that would work here.

Reply:
Sure! Missed that `i` in the loop.